File: 0032-longest-valid-parentheses/0032-longest-valid-parentheses.py
class Solution:
    def longestValidParentheses(self, s: str) -> int:
        openAndClose = [0] * len(s)
        maxLength = 0
        sum = 0
        length = 0

        for i in range(len(openAndClose)):
            openAndClose[i] = 1 if s[i] == '(' else -1
            sum = max(openAndClose[i] + sum, openAndClose[i])
            if sum >= 0:
                length += 1
                if sum == 0:
                    maxLength = max(maxLength, length)
            else:
                length = 0

        sum = 0
        length = 0
        for i in range(len(openAndClose) - 1, -1, -1):
            sum = min(openAndClose[i] + sum, openAndClose[i])
            if sum <= 0:
                length += 1
                if sum == 0:
                    maxLength = max(maxLength, length)
            else:
                length = 0

        return maxLength

File: 0032-longest-valid-parentheses/0032-longest-valid-parentheses.cpp
class Solution {
public:
    int longestValidParentheses(string s) {
        std::stack<int> opening_parenthesises;
        int longest_parentheses = 0, begin_index = 0;;
        for(int i = 0; i < s.length(); i++){
            if(s[i] == '(')
                opening_parenthesises.push(i);
            else if(!(opening_parenthesises.empty())){
                opening_parenthesises.pop();
                if(!(opening_parenthesises.empty()))
                    longest_parentheses = max(longest_parentheses, i - opening_parenthesises.top());
                else
                    longest_parentheses = max(longest_parentheses, i + 1 - begin_index);

            }
            else
                begin_index = i + 1;
        }
        return longest_parentheses;
    }
};